nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,300 @@
+
+    eprop_history_.erase( eprop_history_.begin(), finish ); // erase found entries since no longer used
+  }
+}
+
+void
+nest::EpropArchivingNode::tidy_spike_history()
+{
+  double earliest_time_to_keep = ( t_last_update_per_synapse_.begin() )->t_;
+  while ( ( !spike_history_.empty() ) && ( spike_history_.front() + 1.0e-6 < earliest_time_to_keep ) )
+  {
+    spike_history_.pop_front();
+  }
+}
+
+void
+nest::EpropArchivingNode::write_v_m_pseudo_deriv_to_eprop_history( long time_step, double v_m_pseudo_deriv )
+{
+  if ( n_incoming_ )
+  {
+    const Time time_step_ = Time::step( time_step );
+    const double time_ms = time_step_.get_ms();
+    eprop_history_.push_back( histentry_eprop( time_ms, v_m_pseudo_deriv, 0.0 ) );
+  }
+}
+
+void
+nest::EpropArchivingNode::write_error_signal_to_eprop_history( long time_step, double error_signal )
+{
+  const Time time_step_ = Time::step( time_step );
+  const double time_ms = time_step_.get_ms();
+  eprop_history_.push_back( histentry_eprop( time_ms, 0.0, error_signal ) );
+}
+
+void
+nest::EpropArchivingNode::write_spike_history( long time_step )
+{
+  const Time time_step_ = Time::step( time_step );
+  const double time_ms = time_step_.get_ms();
+  spike_history_.push_back( time_ms );
+}
+
+double
+nest::EpropArchivingNode::get_learning_signal_from_eprop_history( long unsigned int shift )
+{
+  if ( eprop_history_.size() > shift )
+  {
+    return ( ( eprop_history_.rbegin() ) + shift )->learning_signal_;
+  }
+  else
+  {
+    return 0.0;
+  }
+}
+
+void
+nest::EpropArchivingNode::write_learning_signal_to_eprop_history( LearningSignalConnectionEvent& e )
+{
+  const double weight = e.get_weight();
+  const long delay = e.get_delay_steps();
+  const Time stamp = e.get_stamp();
+
+  double t_ms = stamp.get_ms() - 2.0 * Time::get_resolution().get_ms();
+
+  std::deque< histentry_eprop >::iterator start;
+  std::deque< histentry_eprop >::iterator finish;
+
+  // get part of eprop history to which the learning signal is added
+  nest::EpropArchivingNode::get_eprop_history(
+    t_ms, t_ms + Time::delay_steps_to_ms( delay ), &start, &finish ); // implicitely increase access counter
+
+  std::vector< unsigned int >::iterator it = e.begin();
+
+  if ( start != finish && it != e.end() )
+  {
+    double normalized_learning_signal = e.get_coeffvalue( it ); // implicitely decrease access counter
+    start->learning_signal_ += weight * normalized_learning_signal;
+    ++start;
+  }
+}
+
+double
+nest::EpropArchivingNode::calculate_v_m_pseudo_deriv( double V_m, double V_th, double V_th_const ) const
+{
+  // V_th: spiking threshold including adaptive part
+  // V_th_const: constant part of threshold
+  // in LIF neurons without adaptation: V_th = V_th_const
+  double v_scaled = ( V_m - V_th ) / V_th_const;
+  double norm_diff_threshold = 1.0 - std::fabs( v_scaled );
+  double vm_pseudo_deriv = gamma_ * ( norm_diff_threshold > 0.0 ? norm_diff_threshold : 0.0 ) / V_th_const;
+  return vm_pseudo_deriv;
```

Comment:
```suggestion
  const double v_scaled = ( V_m - V_th ) / V_th_const;
  const double norm_diff_threshold = 1.0 - std::fabs( v_scaled );
  const double vm_pseudo_deriv = gamma_ * std::max( 0, norm_diff_threshold ) / V_th_const;
  return vm_pseudo_deriv;
```

Reply:
`std::max` is now fixed. The pseudo derivative function was removed from the `EpropArchivingNode` and moved directly into the update function of the neurons since it's part of the (recurrent) neuron models. This location is also better since now `EpropArchivingNode` solely accommodates functions related to storing to and retrieving from history.